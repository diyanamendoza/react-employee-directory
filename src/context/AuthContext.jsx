import { createContext, useContext, useMemo, useState } from 'react'
import { getUser } from '../services/users'

const AuthContext = createContext()

const AuthProvider = ({ children }) => {
    const currentUser = getUser()
    const [user, setUser] = useState(currentUser ? { id: currentUser.id, email: currentUser.email } : '')
    const value = useMemo(() => ({ user, setUser }), [user])
    return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>
}

const useAuth = () => {
    const context = useContext(AuthContext)
    if (context === undefined) { throw new Error('useAuth not used within Auth Provider')}
    return context
}

export { AuthContext, AuthProvider, useAuth }