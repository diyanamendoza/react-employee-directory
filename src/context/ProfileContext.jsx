import { createContext, useContext, useMemo, useState, useEffect } from 'react'
import { getProfile } from '../services/profiles'

export const ProfileContext = createContext()

export const ProfileProvider = ({ children }) => {
    const [profileStatus, setProfileStatus] = useState(false)
    const [profileData, setProfileData] = useState({}) 

    useEffect(() => {
        const hasProfile = async () => {
            try {
                const profileResult = await getProfile()
                if(profileResult.email) 
                    {
                    setProfileStatus(true)
                    setProfileData(profileResult)
                    }
            } catch (error) { console.log(error) }
        }
        hasProfile()
    }, [])

    const contextValue = useMemo(() => {
        return { profileStatus, profileData, setProfileData }
    }, [profileData])

    return (
        <ProfileContext.Provider value={contextValue}>{children}</ProfileContext.Provider>
    )
}

export const useProfile = () => {
    const context = useContext(ProfileContext)
    if (context === undefined) { throw new Error('useProfile hook not called inside provider')}
    return context
}