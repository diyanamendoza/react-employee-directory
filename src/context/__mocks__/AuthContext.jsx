//MOCK AUTH CONTEXT

import { createContext, useContext, useMemo, useState } from 'react'

const AuthContext = createContext()

const AuthProvider = ({ mockUser, children }) => {
    const [user, setUser] = useState(mockUser ? { id: mockUser.id, email: mockUser.email } : '')
    const value = useMemo(() => ({ user, setUser }), [user])
    return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>
}

const useAuth = () => {
    const context = useContext(AuthContext)
    if (context === undefined) { throw new Error('useAuth not used within Auth Provider')}
    return context
}

export { AuthContext, AuthProvider, useAuth }