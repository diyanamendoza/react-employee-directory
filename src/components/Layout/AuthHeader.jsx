import React from 'react'
import { Link } from 'react-router-dom'
import styles from '../../App.css'
import { useAuth } from '../../context/AuthContext'
import { signOutUser } from '../../services/users'
import { useHistory } from 'react-router-dom'


export default function AuthHeader() {
    const { user, setUser } = useAuth()
    const history = useHistory()

    return (
        <div className={styles.authheader}>
            {user ? (
            <div className={styles.authheaderin}>
            <p>Signed in as {user.email}</p>
            <button type='button' onClick={async () => {
                await signOutUser()
                setUser('')
                history.replace('/')
            }}>Logout</button></div>) 
            : (<Link to='/'><button type='button' aria-label='Sign In'>Sign In/Up</button></Link>) }
        </div>
    )
}
