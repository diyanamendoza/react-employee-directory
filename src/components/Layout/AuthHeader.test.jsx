import { render, screen } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import { AuthProvider } from '../../context/AuthContext'
import AuthHeader from './AuthHeader'

jest.mock('../../context/AuthContext')
jest.mock('../../services/users.js')

it('should display a header with a sign in/up button when logged out', () => {
    const { container } = render(
        <AuthProvider mockUser={null}>
            <MemoryRouter>
                <AuthHeader />
            </MemoryRouter>
        </AuthProvider>
    )

    expect(container).toMatchSnapshot()

    const signinbtn = screen.getByRole('button', { name: 'Sign In'})
    expect(signinbtn).toBeTruthy()
})

it('should display the user email when signed in', () => {
    const { container } = render(
        <AuthProvider mockUser={{ id: 13243, email: 'test@email.com' }}>
            <MemoryRouter>
                <AuthHeader />
            </MemoryRouter>
        </AuthProvider>
    )

    screen.getByText('test@email.com', { exact: false });
    expect(container).toMatchSnapshot();
})