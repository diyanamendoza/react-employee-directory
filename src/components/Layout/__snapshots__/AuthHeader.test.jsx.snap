// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display a header with a sign in/up button when logged out 1`] = `
<div>
  <div
    class="authheader"
  >
    <a
      href="/"
    >
      <button
        aria-label="Sign In"
        type="button"
      >
        Sign In/Up
      </button>
    </a>
  </div>
</div>
`;

exports[`should display the user email when signed in 1`] = `
<div>
  <div
    class="authheader"
  >
    <div
      class="authheaderin"
    >
      <p>
        Signed in as 
        test@email.com
      </p>
      <button
        type="button"
      >
        Logout
      </button>
    </div>
  </div>
</div>
`;
